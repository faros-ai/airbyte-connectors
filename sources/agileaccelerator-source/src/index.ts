import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  Agileaccelerator,
  AgileacceleratorConfig,
} from './agileaccelerator/agileaccelerator';
import {Works} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AgileacceleratorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AgileAccelerator source implementation. */
export class AgileacceleratorSource extends AirbyteSourceBase<AgileacceleratorConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AgileacceleratorConfig
  ): Promise<[boolean, VError]> {
    try {
      const agileaccelerator = await Agileaccelerator.instance(
        config,
        this.logger
      );
      await agileaccelerator.checkConnection(config);
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AgileacceleratorConfig): AirbyteStreamBase[] {
    return [new Works(config, this.logger)];
  }
}
