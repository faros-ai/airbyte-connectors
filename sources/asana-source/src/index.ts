import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {AirbyteSourceLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Asana, AsanaConfig} from './asana';
import {
  Projects,
  ProjectTasks,
  Tags,
  Tasks,
  Users,
  Workspaces,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new AsanaSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Asana source implementation. */
export class AsanaSource extends AirbyteSourceBase<AsanaConfig> {
  get type(): string {
    return 'asana';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AsanaConfig): Promise<[boolean, VError]> {
    try {
      const asana = Asana.instance(config, this.logger);
      await asana.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AsanaConfig): AirbyteStreamBase[] {
    return [
      new Projects(config, this.logger),
      new Tags(config, this.logger),
      new Tasks(config, this.logger),
      new Users(config, this.logger),
      new Workspaces(config, this.logger),
      new ProjectTasks(config, this.logger),
    ];
  }
}
