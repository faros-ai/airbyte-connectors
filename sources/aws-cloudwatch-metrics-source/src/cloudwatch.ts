import {
  CloudWatchClient,
  GetMetricDataCommand,
  GetMetricDataInput,
  GetMetricDataOutput,
  ListMetricsCommand,
  MetricDataQuery,
} from '@aws-sdk/client-cloudwatch';
import {AirbyteLogger} from 'faros-airbyte-cdk';
import _ from 'lodash';
import {VError} from 'verror';

export const MIN_DATE = new Date(0).toISOString();
// January 1, 2200
export const MAX_DATE = new Date(7258118400000).toISOString();

const DEFAULT_CUTOFF_DAYS = 90;
const DEFAULT_PAGE_SIZE = 100;

export interface QueryGroup {
  name: string;
  queries: ReadonlyArray<any>;
}

export interface DataPoint {
  timestamp: string;
  value: number;
  label: string;
}

export interface Config {
  aws_region: string;
  credentials: {
    aws_access_key_id: string;
    aws_secret_access_key: string;
    aws_session_token?: string;
  };
  query_groups: ReadonlyArray<QueryGroup>;
  page_size?: number;
  cutoff_days?: number;
  stream_name?: string;
}

export class CloudWatch {
  private static cloudWatch: CloudWatch;

  constructor(
    private readonly client: CloudWatchClient,
    private readonly startDate: string,
    private readonly endDate: string,
    private readonly pageSize: number
  ) {}

  static instance(config: Config): CloudWatch {
    if (CloudWatch.cloudWatch) return CloudWatch.cloudWatch;

    if (!config.aws_region) {
      throw new VError('Please specify AWS region');
    }
    if (!config.credentials) {
      throw new VError('Please specify AWS credentials');
    }
    if (!config.credentials.aws_access_key_id) {
      throw new VError('Please specify AWS access key ID');
    }
    if (!config.credentials.aws_secret_access_key) {
      throw new VError('Please specify AWS secret access key');
    }

    if (config.query_groups.length === 0) {
      throw new Error('Please specify at least one query group');
    }

    for (const group of config.query_groups) {
      if (!group.name || !group.queries || group.queries.length === 0) {
        throw new Error(
          'Please specify a name and at least one query for each query group'
        );
      }

      for (const query of group.queries) {
        try {
          CloudWatch.toMetricDataQuery(query);
        } catch (e) {
          throw new VError(
            `Query group "${
              group.name
            }" contains invalid query: ${JSON.stringify(query)}`
          );
        }
      }
    }

    const cutoffDays = config.cutoff_days ?? DEFAULT_CUTOFF_DAYS;
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - cutoffDays);
    const startDate = cutoffDate.toISOString();
    const endDate = new Date().toISOString();

    const client = new CloudWatchClient({
      region: config.aws_region,
      credentials: {
        accessKeyId: config.credentials.aws_access_key_id,
        secretAccessKey: config.credentials.aws_secret_access_key,
        sessionToken: config.credentials.aws_session_token,
      },
    });

    CloudWatch.cloudWatch = new CloudWatch(
      client,
      startDate,
      endDate,
      config.page_size ?? DEFAULT_PAGE_SIZE
    );

    return CloudWatch.cloudWatch;
  }

  async checkConnection(): Promise<void> {
    await this.client.send(new ListMetricsCommand({}));
  }

  async *getMetricData(
    queries: ReadonlyArray<any>,
    after?: string,
    logger?: AirbyteLogger
  ): AsyncGenerator<DataPoint> {
    const params: GetMetricDataInput = {
      StartTime: new Date(after ?? this.startDate),
      EndTime: new Date(this.endDate),
      MaxDatapoints: this.pageSize,
      MetricDataQueries: queries.map(CloudWatch.toMetricDataQuery),
    };

    do {
      const command = new GetMetricDataCommand(params);
      const response: GetMetricDataOutput = await this.client.send(command);

      for (const result of response.MetricDataResults ?? []) {
        for (const [t, v] of _.zip(result.Timestamps, result.Values)) {
          if (t === undefined || v === undefined) continue;
          yield {
            timestamp: t.toISOString(),
            value: v,
            label: result.Label,
          };
        }
        logger?.info(`Fetched ${result.Timestamps?.length} data points`);
      }

      params.NextToken = response?.NextToken;
    } while (params.NextToken);
  }

  static toMetricDataQuery(query: any): MetricDataQuery | undefined {
    if (typeof query === 'object') {
      return query;
    }

    if (typeof query === 'string') {
      return JSON.parse(query);
    }

    return undefined;
  }

  static toQueryString(query: any): string | undefined {
    if (typeof query === 'string') {
      return query;
    }

    if (typeof query === 'object') {
      return JSON.stringify(query);
    }

    return undefined;
  }
}
