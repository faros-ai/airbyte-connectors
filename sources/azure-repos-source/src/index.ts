import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureRepoConfig, AzureRepos} from './azure-repos';
import {PullRequests, Repositories, Users} from './streams';
import {Commits} from './streams/commits';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureRepoSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzureRepo source implementation. */
export class AzureRepoSource extends AirbyteSourceBase<AzureRepoConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AzureRepoConfig): Promise<[boolean, VError]> {
    try {
      const azureRepos = await AzureRepos.make(config, this.logger);
      await azureRepos.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureRepoConfig): AirbyteStreamBase[] {
    return [
      new Commits(config, this.logger),
      new PullRequests(config, this.logger),
      new Repositories(config, this.logger),
      new Users(config, this.logger),
    ];
  }
}
