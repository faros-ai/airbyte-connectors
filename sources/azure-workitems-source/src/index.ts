import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureWorkitems, AzureWorkitemsConfig} from './azure-workitems';
import {Boards, Iterations, Users} from './streams';
import {Workitems} from './streams/workitems';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureWorkitemsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Example source implementation. */
export class AzureWorkitemsSource extends AirbyteSourceBase<AzureWorkitemsConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AzureWorkitemsConfig
  ): Promise<[boolean, VError]> {
    try {
      const azureActiveDirectory = await AzureWorkitems.instance(config);
      await azureActiveDirectory.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureWorkitemsConfig): AirbyteStreamBase[] {
    return [
      new Workitems(config, this.logger),
      new Users(config, this.logger),
      new Iterations(config, this.logger),
      new Boards(config, this.logger),
    ];
  }
}
