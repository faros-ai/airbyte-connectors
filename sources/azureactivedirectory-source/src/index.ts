import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  AzureActiveDirectory,
  AzureActiveDirectoryConfig,
} from './azureactivedirectory';
import {Groups, Users} from './streams';
/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureActiveDirectorySource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzureActiveDirectory source implementation. */
export class AzureActiveDirectorySource extends AirbyteSourceBase<AzureActiveDirectoryConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AzureActiveDirectoryConfig
  ): Promise<[boolean, VError]> {
    try {
      const azureActiveDirectory = await AzureActiveDirectory.instance(
        config,
        this.logger
      );
      await azureActiveDirectory.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureActiveDirectoryConfig): AirbyteStreamBase[] {
    return [new Users(config, this.logger), new Groups(config, this.logger)];
  }
}
