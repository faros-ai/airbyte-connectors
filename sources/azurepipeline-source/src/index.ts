import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzurePipeline, AzurePipelineConfig} from './azurepipeline';
import {Builds, Pipelines, Releases} from './streams';
/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzurePipelineSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzurePipeline source implementation. */
export class AzurePipelineSource extends AirbyteSourceBase<AzurePipelineConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AzurePipelineConfig
  ): Promise<[boolean, VError]> {
    try {
      const azureActiveDirectory = AzurePipeline.instance(config);
      await azureActiveDirectory.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzurePipelineConfig): AirbyteStreamBase[] {
    return [
      new Pipelines(config, this.logger),
      new Builds(config, this.logger),
      new Releases(config, this.logger),
    ];
  }
}
