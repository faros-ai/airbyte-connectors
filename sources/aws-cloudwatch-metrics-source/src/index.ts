import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {CloudWatch, Config} from './cloudwatch';
import {Metrics} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new CloudWatchMetricsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** CloudWatch Metrics source implementation. */
export class CloudWatchMetricsSource extends AirbyteSourceBase<Config> {
  get type(): string {
    return 'cloudwatch-metrics';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: Config): Promise<[boolean, VError]> {
    try {
      const cloudWatch = CloudWatch.instance(config);
      await cloudWatch.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: Config): AirbyteStreamBase[] {
    return [new Metrics(config, this.logger)];
  }
}
