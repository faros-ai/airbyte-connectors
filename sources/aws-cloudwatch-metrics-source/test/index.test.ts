import {
  AirbyteLogLevel,
  AirbyteSourceLogger,
  AirbyteSpec,
  SyncMode,
} from 'faros-airbyte-cdk';
import fs from 'fs-extra';
import VError from 'verror';

import {CloudWatch, Config} from '../src/cloudwatch';
import * as sut from '../src/index';

describe('index', () => {
  const logger = new AirbyteSourceLogger(
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.FATAL
  );

  const validConfig: Config = {
    aws_region: 'us-west-2',
    credentials: {
      aws_access_key_id: 'YOUR_AWS_ACCESS_KEY_ID',
      aws_secret_access_key: 'YOUR_AWS_SECRET_ACCESS_KEY',
      aws_session_token: 'YOUR_AWS_SESSION_TOKEN',
    },
    query_groups: [
      {
        name: 'ExampleQueryGroup',
        queries: [{query: '{"metric":"CPUUtilization"}'}],
      },
    ],
    page_size: 100,
  };

  const source = new sut.CloudWatchMetricsSource(logger);

  test('spec', async () => {
    await expect(source.spec()).resolves.toStrictEqual(
      new AirbyteSpec(readResourceFile('spec.json'))
    );
  });

  test('check connection - invalid', async () => {
    await expect(
      source.checkConnection({...validConfig, aws_region: undefined})
    ).resolves.toStrictEqual([false, new VError('Please specify AWS region')]);
    await expect(
      source.checkConnection({...validConfig, credentials: undefined})
    ).resolves.toStrictEqual([
      false,
      new VError('Please specify AWS credentials'),
    ]);
    await expect(
      source.checkConnection({
        ...validConfig,
        credentials: {
          aws_access_key_id: undefined,
          aws_secret_access_key: 'YOUR_AWS_SECRET_ACCESS_KEY',
        },
      })
    ).resolves.toStrictEqual([
      false,
      new VError('Please specify AWS access key ID'),
    ]);
    await expect(
      source.checkConnection({
        ...validConfig,
        credentials: {
          aws_access_key_id: 'YOUR_AWS_ACCESS_KEY_ID',
          aws_secret_access_key: undefined,
        },
      })
    ).resolves.toStrictEqual([
      false,
      new VError('Please specify AWS secret access key'),
    ]);
  });

  test('check connection', async () => {
    CloudWatch.instance = jest.fn().mockImplementation(() => {
      return new CloudWatch(
        {
          send: jest.fn().mockResolvedValue({}),
        } as any,
        '2021-01-01',
        '2021-01-02',
        100
      );
    });

    await expect(source.checkConnection(validConfig)).resolves.toStrictEqual([
      true,
      undefined,
    ]);
  });

  test('metrics incremental', async () => {
    const fnSend = jest.fn();
    const latestTimestamp = '2021-01-02T00:00:00.000Z';
    const expectedData = {
      MetricDataResults: [
        {
          Id: 'm1',
          Label: 'CPUUtilization',
          Timestamps: [
            new Date('2021-01-01T00:00:00.000Z'),
            new Date(latestTimestamp),
          ],
          Values: [0.5, 0.6],
        },
      ],
    };
    CloudWatch.instance = jest.fn().mockImplementation(() => {
      return new CloudWatch(
        {
          send: fnSend.mockResolvedValue(expectedData),
        } as any,
        '2021-01-01',
        '2021-01-02',
        100
      );
    });

    const streams = source.streams(validConfig);
    const stream = streams[0];
    const iter = stream.readRecords(
      SyncMode.INCREMENTAL,
      undefined,
      {queryGroup: validConfig.query_groups[0], queryHash: 'hash'},
      {
        ExampleQueryGroup: {
          hash: {
            timestamp: '2021-01-01T00:00:00.000Z',
          },
        },
      }
    );

    const items = [];
    for await (const item of iter) {
      items.push(item);
    }

    expect(fnSend).toHaveBeenCalledTimes(1);
    expect(items).toMatchSnapshot();
    expect(stream.getUpdatedState(undefined, undefined)).toEqual({
      ExampleQueryGroup: {
        hash: {
          timestamp: latestTimestamp,
        },
      },
    });
  });

  test('metrics full', async () => {
    const fnSend = jest.fn();
    const expectedData = {
      MetricDataResults: [
        {
          Id: 'm1',
          Label: 'CPUUtilization',
          Timestamps: [
            new Date('2021-01-01T00:00:00.000Z'),
            new Date('2021-01-02T00:00:00.000Z'),
          ],
          Values: [0.5, 0.6],
        },
      ],
    };
    CloudWatch.instance = jest.fn().mockImplementation(() => {
      return new CloudWatch(
        {
          send: fnSend.mockResolvedValue(expectedData),
        } as any,
        '2021-01-01',
        '2021-01-02',
        100
      );
    });

    const streams = source.streams(validConfig);
    const stream = streams[0];
    const iter = stream.readRecords(SyncMode.FULL_REFRESH, undefined, {
      queryGroup: validConfig.query_groups[0],
      queryHash: 'hash',
    });

    const items = [];
    for await (const item of iter) {
      items.push(item);
    }

    expect(fnSend).toHaveBeenCalledTimes(1);
    expect(items).toMatchSnapshot();
  });
});

function readResourceFile(fileName: string): any {
  return JSON.parse(fs.readFileSync(`resources/${fileName}`, 'utf8'));
}
