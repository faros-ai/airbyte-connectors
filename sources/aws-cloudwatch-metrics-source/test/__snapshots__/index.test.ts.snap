// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index metrics full 1`] = `
Array [
  Object {
    "label": "CPUUtilization",
    "queryName": "ExampleQueryGroup",
    "timestamp": "2021-01-01T00:00:00.000Z",
    "value": 0.5,
  },
  Object {
    "label": "CPUUtilization",
    "queryName": "ExampleQueryGroup",
    "timestamp": "2021-01-02T00:00:00.000Z",
    "value": 0.6,
  },
]
`;

exports[`index metrics incremental 1`] = `
Array [
  Object {
    "label": "CPUUtilization",
    "queryName": "ExampleQueryGroup",
    "timestamp": "2021-01-01T00:00:00.000Z",
    "value": 0.5,
  },
  Object {
    "label": "CPUUtilization",
    "queryName": "ExampleQueryGroup",
    "timestamp": "2021-01-02T00:00:00.000Z",
    "value": 0.6,
  },
]
`;
